axum/src/routing/tests/mod.rs

```diff
@@ -799,3 +803,23 @@ async fn layer_response_into_response() {
     assert_eq!(res.headers()["x-foo"], "bar");
     assert_eq!(res.status(), StatusCode::IM_A_TEAPOT);
 }
+
+// https://github.com/tokio-rs/axum/issues/1579
+#[tokio::test]
+async fn routing_to_router() {
+    let inner = Router::new().route(
+        "/foobar",
+        get(|matched_path: MatchedPath| async move { matched_path.as_str().to_owned() }),
+    );
+
+    let app = Router::new().route(
+        "/*path",
+        any(|req| Router::new().nest("/", inner).oneshot(req)),
+    );
+
+    let client = TestClient::new(app);
+
+    let res = client.get("/foobar").send().await;
+    assert_eq!(res.status(), StatusCode::OK);
+    assert_eq!(res.text().await, "/*path/foobar");
```

Comment:
I'm not really sure what the matched path logically should be. Feels weird saying it is `/*path/foobar` since the request is just `/foobar`, so where does `/*path` come from?

However it kinda matches the way the router is setup 🤔

Reply:
cc @Pure-Peace

Reply:
In general I think a more correct router setup would be

```rust
async fn handler(req: Request<Body>) -> crate::response::Response {
    let inner = Router::new().route(
        "/foobar",
        get(|matched_path: MatchedPath| async move { matched_path.as_str().to_owned() }),
    );

    Router::new().nest("/", inner).oneshot(req).await.unwrap()
}

// run it with
handler.into_make_service();
```

Reply:
generally like this

I'm using axum to make some framework like

```rust
pub struct App {
    pub cfg: Arc<GatewayConfig>,
}

impl Application for App {
    fn frame_cfg(&self) -> &ApiFrameConfig {
        &self.cfg.frame_cfg
    }

    fn router(&self) -> Router {
        Router::new()
            .route("/", get(peace_api::responder::app_root))
            .nest("/bancho", bancho::routers::bancho_client_routes())
    }

    fn apidocs(&self) -> utoipa::openapi::OpenApi {
        GatewayApiDocs::openapi()
    }

    fn match_hostname(
        &self,
        Host(hostname): Host,
        req: &Request<Body>,
    ) -> Option<Router> {
        match hostname {
            n if self.cfg.bancho_hostname.contains(&n) => {
                Some(bancho::routers::bancho_client_routes())
            },
            _ => None,
        }
    }
}

#[tokio::main]
pub async fn main() {
    let app = App::new(GatewayConfig::get());
    http::serve(app).await;
}
```

In general a reverse proxy can be used to direct the request to a certain path, which may be the router of a certain service

I would like to allow users to enable both `general router` and `hostname based router`. This may need to use `/*path`

possible to reuse some routers